src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts

```diff
@@ -1,38 +1,102 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+import { ComponentDescriptor, ServerComponentDescriptor, WebAssemblyComponentDescriptor, discoverComponents } from '../../Services/ComponentDescriptorDiscovery';
+import { isInteractiveRootComponentElement } from '../BrowserRenderer';
 import { applyAnyDeferredValue } from '../DomSpecialPropertyUtil';
+import { LogicalElement, getLogicalChildrenArray, getLogicalNextSibling, getLogicalParent, getLogicalRootDescriptor, insertLogicalChild, insertLogicalChildBefore, isLogicalElement, toLogicalElement, toLogicalRootCommentElement } from '../LogicalElements';
 import { synchronizeAttributes } from './AttributeSync';
 import { UpdateCost, ItemList, Operation, computeEditScript } from './EditScript';
 
-export function synchronizeDomContent(destination: CommentBoundedRange | Node, newContent: Node) {
-  let destinationParent: Node;
-  let nextDestinationNode: Node | null;
-  let originalNodesForDiff: ItemList<Node>;
+let descriptorHandler: DescriptorHandler | null = null;
+
+export interface DescriptorHandler {
+  registerComponentDescriptor(descriptor: ComponentDescriptor): void;
+}
+
+export function attachComponentDescriptorHandler(handler: DescriptorHandler) {
+  descriptorHandler = handler;
+}
+
+export function registerAllComponentDescriptors(root: Node) {
+  const descriptors = upgradeComponentCommentsToLogicalRootComments(root);
 
-  // Figure out how to interpret the 'destination' parameter, since it can come in two very different forms
+  for (const descriptor of descriptors) {
+    descriptorHandler?.registerComponentDescriptor(descriptor);
+  }
+}
+
+export { preprocessAndSynchronizeDomContent as synchronizeDomContent };
+
+function preprocessAndSynchronizeDomContent(destination: CommentBoundedRange | Node, newContent: Node) {
+  // Start by recursively identifying component markers in the new content
+  // and converting them into logical elements so they correctly participate
+  // in logical element synchronization
+  upgradeComponentCommentsToLogicalRootComments(newContent);
+
+  // Then, synchronize the preprocessed DOM content
+  synchronizeDomContentCore(destination, newContent);
+}
+
+function synchronizeDomContentCore(destination: CommentBoundedRange | Node, newContent: Node) {
```

Comment:
Entirely mixing together the DOM sync algorithm with special-case knowledge of logical elements representation and the interactive component markers is not ideal and will make this pretty hard to understand and maintain in the future. DOM syncing is already fairly sophisticated so it would be extremely valuable to layer this in a way that the interactive component marker handling and logical element walking is handled externally to this code.

Is there a way of adding a bit more abstraction to `DomSync.ts` so that it doesn't need to know about interactive component markers and the difference between logical and physical elements? What you have with `EditWalker` is a good start - maybe the `synchronizeDomContent` export itself could be defined to receive some abstraction like that as the input and output data.

There is good test coverage for the existing `DomSync.ts` behavior in `src\Components\Web.JS\test\DomSync.test.ts`, so anything we add/change in `DomSync` should be reflected in unit tests. The ultimate ideal would be if it didn't really gain new behavior at all besides accepting a more general abstraction to represent the DOM structures it's working over!

What do you think? Or is it impractical to add an abstraction that's still able to deal with some necessary special cases?